R/autosyn.R

```diff
@@ -121,17 +108,28 @@ flywire_partners <- function(rootid, partners=c("outputs", "inputs"),
     colnames(resdf) <- c("offset", 'pre_svid', "post_svid", "scores", "cleft_scores")
     # we can get the same row appearing twice for autapses
     resdf <- filter(resdf, !duplicated(.data$offset))
-    resdf <- if(partners=='outputs')
-      filter(resdf, .data$pre_svid%in%svids) else filter(resdf, .data$post_svid%in%svids)
-
+    if (partners == "outputs"){
+      resdf <-  filter(resdf, .data$pre_svid %in% svids)
+    }else if (partners == "inputs"){
+      resdf <-  filter(resdf, .data$post_svid %in% svids)
+    }
   } else {
-    df <- if(partners=="outputs")
-      tibble::tibble(pre_svid=svids) else tibble::tibble(post_svid=svids)
-    res=dplyr::inner_join(flywireids, df,
-                          by=ifelse(partners=="outputs", "pre_svid", "post_svid"),
-                          copy=TRUE, auto_index=TRUE, )
+    if(partners == "both"){
+      res = dplyr::filter(flywireids,
+                          .data$pre_svid%in%svids|.data$post_svid%in%svids)
```

Comment:
@alexanderbates you mentioned that most of the time was spent in sqlite queries. With your current code, try doing 'both' and then compare with doing outputs and inputs separately. You will find that the latter is *much* quicker. Your query does not use the indexes on pre_svid and post_svid and intead does a full table scan. Furthermore it embeds potentially hundreds of thousands of ids into a string SQL query, which is also suboptimal. The query strategy in my original code uses an indexed inner join on a temporary table. This may be 100x faster.

Reply:
Here is an example for a PN:
```
> bench::mark(both=flywire_partners("720575940615237849", partners = 'both', Verbose = F), ins=flywire_partners("720575940615237849", partners = 'in', Verbose = F), out=flywire_partners("720575940615237849", partners = 'out', Verbose = F), check = F)
Fetching root ids
Fetching root ids
Fetching root ids
# A tibble: 3 x 13
  expression   min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
  <bch:expr> <bch> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>
1 both        1.6m   1.6m    0.0104   52.71MB   0.0104     1     1
2 ins        2.61s  2.61s    0.383     4.63MB   0          1     0
3 out        4.22s  4.22s    0.237     7.14MB   0.237      1     1
```
this take ~100s vs 6s and the differential increases as the neuron size increases.

Reply:
Okay good point - I am not that used to using SQL databases in R. I can't see the tidying up you did with me last night on the remote. You want me to commit further changes to this branch or wait for you to push the tidying you did (either published as separate branches or an update to this one?)

Reply:
Should we be calling dbDisconnect() when we finish working with the .db

Reply:
Greg says:

https://github.com/natverse/fafbseg/pull/73 merged
https://github.com/natverse/fafbseg/pull/77 pending your review but I think good to merge now
I will put up the last PR now.
Note that there was a significant issue in 77 which was resulting in very slow sqlite queries – the final as.data.frame step triggering the sqlite query to run quickly got extremely slow for large neurons.

Reply:
> Should we be calling dbDisconnect() when we finish working with the .db

I am reusing the same *read only* sqlite connection (by memoising) the dbplyr tbl objects (see `memo_sqlite_con()` and `memo_tbl()`) and not closing it across queries. I think this is unlikely to cause any trouble, but if you find any clear advice to the contrary I would be interested.